src/Features/Core/Portable/UpgradeProject/AbstractUpgradeProjectCodeFixProvider.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace Microsoft.CodeAnalysis.UpgradeProject
+{
+    internal abstract partial class AbstractUpgradeProjectCodeFixProvider : CodeFixProvider
+    {
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
```

Comment:
the core algorithm could be here (esp. as it will be the same for VB).  Plus this is where you can put your SolutoinChangeAction subclass.  Only keep C# specific logic in the C# layer.